File: py-recitations-solutions/challenge/roman.py
# roman.py

# https://en.wikipedia.org/wiki/Roman_numerals

def romannumeral(n:int)->str:
    '''
    Limited to n = 3999 due to symbolic changes past this number
    '''
    #   Write a program to turn an integer into its corresponding Roman numeral
    #   representation.
    if  n > 3999 or n <= 0:
        print("Roman Numeral Representation limited to NATURAL NUMBERS under 3999.")
        return None

    r = ""

    # Roman Numerals
    M = 1000
    # D = 500
    C = 100
    # L = 50
    X = 10
    # V = 5
    I = 1

    r1000 = int((n - n%(M))/M)
    r100 = int((n%(M) - n%(C))/C)
    r10 = int((n%(C) - n%(X))/X)
    r1 = int((n%(X) - n%I)/I)

    # 1000s
    for i in range(r1000):
        r = r + "M"
    
    # Dealing with cases for each digit

    # 100s
    if r100 == 9:
        r =  r + "CM"
    elif 5 <= r100 < 9:
        r = r + "D"
        if r100 - 5 > 0:
            for i in range(r100-5):
                r = r + "C"
    elif r100 == 4:
        r =  r + "CD"
    else:
        for i in range(r100):
            r = r + "C"

    # 10s
    if r10 == 9:
        r =  r + "XC"
    elif 5 <= r10 < 9:
        r = r + "L"
        if r10 - 5 > 0:
            for i in range(r10-5):
                r = r + "X"
    elif r10 == 4:
        r =  r + "XL"
    else:
        for i in range(r10):
            r = r + "X"

    # 1s
    if r1 == 9:
        r =  r + "IX"
    elif 5 <= r1 < 9:
        r = r + "V"
        if r1 - 5 > 0:
            for i in range(r1-5):
                r = r + "I"
    elif r1 == 4:
        r =  r + "IV"
    else:
        for i in range(r1):
            r = r + "I"

    return r

if __name__ == "__main__":

    year = 2017

    # What is your birth year in roman numerals?
    # Output result to user
    if romannumeral(year) != None:
        print(f"{year} in Roman Numerals is : {romannumeral(year)}")

File: cpp-recitations-solutions/challenge/roman.cpp
#include <iostream>
using namespace std;

void romannumeral(int n){
  // Limited to n = 3999 due to symbolic changes past this number
  string output = " ";

  if(n>3999 || n<1) cout<< n << " is out of range, upper bound is 3999, lower bound is 1"; 
  else{
  // Roman Numerals
  int M,C,X,I,rM,rC,rX,rI;
  M = 1000;
  C = 100;
  X = 10;
  I = 1;

  rM = (n - n%M)/M;
  rC = (n%M-n%C)/C;
  rX = (n%C-n%X)/X;
  rI = (n%X-n%I)/I;

  // 1000s
  for(int i = 0; i < rM; i++) output = output + "M";
  // 100s
  if(rC==9) {output = output + "CM";}
  else if(rC < 9 && rC>=5) {output = output+"D"; if(rC-5>0){for(int i = 0; i < rC-5; i++){output = output + "C";}}}
  else if(rC == 4) {output = output + "CD";}
  else {for(int i = 0; i < rC; i++){output = output + "C";}}
  // 10s
  if(rX==9) {output = output + "XC";}
  else if(rX < 9 && rX>=5) {output = output+"L"; if(rX-5>0){for(int i = 0; i < rX-5; i++){output = output + "X";}}}
  else if(rX == 4) {output = output + "XL";}
  else {for(int i = 0; i < rX; i++){output = output + "X";}}
  // 1s
  if(rI==9) {output = output + "IX";}
  else if(rI < 9 && rI>=5) {output = output+"V"; if(rI-5>0){for(int i = 0; i < rI-5; i++){output = output + "I";}}}
  else if(rI == 4) {output = output + "IV";}
  else {for(int i = 0; i < rI; i++){output = output + "I";}}

  cout << n << " in Roman Numerals is "<< output;
  }
}

int main() {
  
  // Write a program to turn an integer into its corresponding Roman numeral
  // representation.
  // https://en.wikipedia.org/wiki/Roman_numerals

  int year;
  cout<<"Input a year : ";
  cin >> year;
  romannumeral(year);  
}